From e2d1e4b05f0f753ed16be6734f5a4c9666e5fce5 Mon Sep 17 00:00:00 2001
From: Xindong Xu <xindong.xu@amlogic.com>
Date: Wed, 3 Nov 2021 16:30:13 +0800
Subject: [PATCH] vold: fix format storage as Device Storage error [1/1]

PD#SWPL-61415

Problem:
format storage as Device Storage error

Solution:
1. pass size to mke2fs
2. add sleep

Verify:
ohm

Change-Id: I91eee19ca6454185e2a723f807b6494af4175813
---
 Utils.cpp               | 1 +
 model/PrivateVolume.cpp | 8 +++++++-
 2 files changed, 8 insertions(+), 1 deletion(-)
 mode change 100644 => 100755 Utils.cpp

diff --git a/Utils.cpp b/Utils.cpp
old mode 100644
new mode 100755
index 17921e8e..ff051710
--- a/Utils.cpp
+++ b/Utils.cpp
@@ -745,6 +745,7 @@ status_t ForkExecvp(const std::vector<std::string>& args, std::vector<std::strin
         LOG(ERROR) << "Process exited with code: " << WEXITSTATUS(status);
         return WEXITSTATUS(status);
     }
+    sleep(1);
     return OK;
 }
 
diff --git a/model/PrivateVolume.cpp b/model/PrivateVolume.cpp
index b16db5f9..3d3777de 100644
--- a/model/PrivateVolume.cpp
+++ b/model/PrivateVolume.cpp
@@ -211,6 +211,8 @@ status_t PrivateVolume::doUnmount() {
 
 status_t PrivateVolume::doFormat(const std::string& fsType) {
     std::string resolvedFsType = fsType;
+    uint64_t mSize = 0;
+
     if (fsType == "auto") {
         // For now, assume that all MMC devices are flash-based SD cards, and
         // give everyone else ext4 because sysfs rotational isn't reliable.
@@ -236,7 +238,11 @@ status_t PrivateVolume::doFormat(const std::string& fsType) {
             }
         }
         // TODO: change reported mountpoint once we have better selinux support
-        if (ext4::Format(mDmDevPath, 0, "/data")) {
+        if (GetBlockDevSize(mRawDevPath, &mSize) != OK) {
+            mSize = 0;
+        }
+
+        if (ext4::Format(mDmDevPath, (mSize / 32768), "/data")) {
             PLOG(ERROR) << getId() << " failed to format";
             return -EIO;
         }
-- 
2.28.0

